---
[v3] net: ethernet: mtk_eth_soc: use generic allocator for MT7987

This patch introduces a fallback to the DRAM path to the existing
generic allocator patch when the kernel lacks sufficient SRAM regions
for allocating TX/RX/FQ rings. This enables MTK_SRAM feature on
platforms with insufficient SRAM, improving compatibility across
different SoCs.
---
v3: Use generic allocator method to instead of mtk_validate_sram_range(),
    and change patch name from '999-eth-22-mtk_eth_soc-add-mtk_validate_
    sram_range.patch'
v2: Fix the ADMA RX hang issue that occurs during the reboot test.
v1: Add mtk_validate_sram_range() to the MTK_SRAM feature.
---

From 7fb965b100f4b5f70ea1ac7e0d05dfb8b0a78931 Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Mon, 19 Jan 2026 16:41:44 +0800
Subject: [PATCH] net: ethernet: mtk_eth_soc: use generic allocator for
 MT7987

This patch introduces a fallback to the DRAM path to the existing
generic allocator patch when the kernel lacks sufficient SRAM regions
for allocating TX/RX/FQ rings. This enables MTK_SRAM feature on
platforms with insufficient SRAM, improving compatibility across
different SoCs.

Signed-off-by: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c | 93 +++++++++++++--------
 drivers/net/ethernet/mediatek/mtk_eth_soc.h | 24 +++--
 2 files changed, 77 insertions(+), 40 deletions(-)

--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -1394,19 +1394,23 @@ static void *mtk_max_lro_buf_alloc(gfp_t
 }
 
 static void *mtk_dma_ring_alloc(struct mtk_eth *eth, size_t size,
-				dma_addr_t *dma_handle, bool use_sram)
+				dma_addr_t *dma_handle, bool use_sram, bool *in_sram)
 {
-	void *dma_ring;
+	void *dma_ring = NULL;
 
 	if (use_sram && eth->sram_pool) {
 		dma_ring = (void *)gen_pool_alloc(eth->sram_pool, size);
-		if (!dma_ring)
-			return dma_ring;
-		*dma_handle = gen_pool_virt_to_phys(eth->sram_pool,
-						    (unsigned long)dma_ring);
-	} else {
+		if (dma_ring) {
+			*dma_handle = gen_pool_virt_to_phys(eth->sram_pool,
+							    (unsigned long)dma_ring);
+			*in_sram = true;
+		}
+	}
+
+	if (!dma_ring) {
 		dma_ring = dma_alloc_coherent(eth->dma_dev, size, dma_handle,
 					      GFP_KERNEL);
+		*in_sram = false;
 	}
 
 	return dma_ring;
@@ -1430,23 +1434,24 @@ static int mtk_init_fq_dma(struct mtk_et
 	dma_addr_t dma_addr;
 	int i, j, len;
 
-	eth->scratch_ring = mtk_dma_ring_alloc(eth, TX_DESC_OFS(eth, cnt),
-					       &eth->phy_scratch_ring, true);
+	eth->fq_ring.scratch_ring = mtk_dma_ring_alloc(eth, TX_DESC_OFS(eth, cnt),
+						       &eth->fq_ring.phy_scratch_ring, true,
+						       &eth->fq_ring.in_sram);
 
-	if (unlikely(!eth->scratch_ring))
+	if (unlikely(!eth->fq_ring.scratch_ring))
 		return -ENOMEM;
 
-	phy_ring_tail = eth->phy_scratch_ring + TX_DESC_OFS(eth, cnt - 1);
+	phy_ring_tail = eth->fq_ring.phy_scratch_ring + TX_DESC_OFS(eth, cnt - 1);
 
 	for (j = 0; j < DIV_ROUND_UP(soc->tx.fq_dma_size, MTK_FQ_DMA_LENGTH); j++) {
 		len = min_t(int, cnt - j * MTK_FQ_DMA_LENGTH, MTK_FQ_DMA_LENGTH);
-		eth->scratch_head[j] = kcalloc(len, MTK_QDMA_PAGE_SIZE, GFP_KERNEL);
+		eth->fq_ring.scratch_head[j] = kcalloc(len, MTK_QDMA_PAGE_SIZE, GFP_KERNEL);
 
-		if (unlikely(!eth->scratch_head[j]))
+		if (unlikely(!eth->fq_ring.scratch_head[j]))
 			return -ENOMEM;
 
 		dma_addr = dma_map_single(eth->dma_dev,
-					  eth->scratch_head[j], len * MTK_QDMA_PAGE_SIZE,
+					  eth->fq_ring.scratch_head[j], len * MTK_QDMA_PAGE_SIZE,
 					  DMA_FROM_DEVICE);
 
 		if (unlikely(dma_mapping_error(eth->dma_dev, dma_addr)))
@@ -1455,10 +1460,10 @@ static int mtk_init_fq_dma(struct mtk_et
 		for (i = 0; i < len; i++) {
 			struct mtk_tx_dma_v2 *txd;
 
-			txd = eth->scratch_ring + TX_DESC_OFS(eth, j * MTK_FQ_DMA_LENGTH + i);
+			txd = eth->fq_ring.scratch_ring + TX_DESC_OFS(eth, j * MTK_FQ_DMA_LENGTH + i);
 			txd->txd1 = dma_addr + i * MTK_QDMA_PAGE_SIZE;
 			if (j * MTK_FQ_DMA_LENGTH + i < cnt)
-				txd->txd2 = eth->phy_scratch_ring +
+				txd->txd2 = eth->fq_ring.phy_scratch_ring +
 					    TX_DESC_OFS(eth, j * MTK_FQ_DMA_LENGTH + i + 1);
 
 			txd->txd3 = TX_DMA_PLEN0(MTK_QDMA_PAGE_SIZE);
@@ -1475,7 +1480,7 @@ static int mtk_init_fq_dma(struct mtk_et
 		}
 	}
 
-	mtk_w32(eth, eth->phy_scratch_ring, soc->reg_map->qdma.fq_head);
+	mtk_w32(eth, eth->fq_ring.phy_scratch_ring, soc->reg_map->qdma.fq_head);
 	mtk_w32(eth, phy_ring_tail, soc->reg_map->qdma.fq_tail);
 	mtk_w32(eth, (cnt << 16) | cnt, soc->reg_map->qdma.fq_count);
 	mtk_w32(eth, MTK_QDMA_PAGE_SIZE << 16, soc->reg_map->qdma.fq_blen);
@@ -2914,7 +2919,8 @@ static int mtk_tx_alloc(struct mtk_eth *
 	if (!ring->buf)
 		goto no_tx_mem;
 
-	ring->dma = mtk_dma_ring_alloc(eth, ring_size * sz, &ring->phys, true);
+	ring->dma = mtk_dma_ring_alloc(eth, ring_size * sz, &ring->phys, true,
+				       &ring->in_sram);
 	if (!ring->dma)
 		goto no_tx_mem;
 
@@ -3021,7 +3027,7 @@ static void mtk_tx_clean(struct mtk_eth
 
 	if (ring->dma) {
 		mtk_dma_ring_free(eth, TX_DESC_OFS(eth, ring->dma_size),
-				  ring->dma, ring->phys, true);
+				  ring->dma, ring->phys, ring->in_sram);
 		ring->dma = NULL;
 	}
 
@@ -3078,7 +3084,8 @@ static int mtk_rx_alloc(struct mtk_eth *
 	ring->dma = mtk_dma_ring_alloc(eth,
 				       RX_DESC_OFS(eth, rx_dma_size),
 				       &ring->phys,
-				       rx_flag == MTK_RX_FLAGS_NORMAL);
+				       rx_flag == MTK_RX_FLAGS_NORMAL,
+				       &ring->in_sram);
 	if (!ring->dma)
 		return -ENOMEM;
 
@@ -3923,15 +3930,15 @@ static void mtk_dma_free(struct mtk_eth
 			netdev_tx_reset_subqueue(eth->netdev[i], j);
 	}
 
-	if (eth->scratch_ring) {
+	if (eth->fq_ring.scratch_ring) {
 		mtk_dma_ring_free(eth, TX_DESC_OFS(eth, soc->tx.fq_dma_size),
-				  eth->scratch_ring, eth->phy_scratch_ring,
-				  true);
-		eth->scratch_ring = NULL;
-		eth->phy_scratch_ring = 0;
+				  eth->fq_ring.scratch_ring, eth->fq_ring.phy_scratch_ring,
+				  eth->fq_ring.in_sram);
+		eth->fq_ring.scratch_ring = NULL;
+		eth->fq_ring.phy_scratch_ring = 0;
 	}
 	mtk_tx_clean(eth);
-	mtk_rx_clean(eth, &eth->rx_ring[0], true);
+	mtk_rx_clean(eth, &eth->rx_ring[0], eth->rx_ring[0].in_sram);
 	mtk_rx_clean(eth, &eth->rx_ring_qdma, false);
 
 	if (eth->hwlro) {
@@ -3948,8 +3955,8 @@ static void mtk_dma_free(struct mtk_eth
 	}
 
 	for (i = 0; i < DIV_ROUND_UP(soc->tx.fq_dma_size, MTK_FQ_DMA_LENGTH); i++) {
-		kfree(eth->scratch_head[i]);
-		eth->scratch_head[i] = NULL;
+		kfree(eth->fq_ring.scratch_head[i]);
+		eth->fq_ring.scratch_head[i] = NULL;
 	}
 }
 
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -1110,6 +1110,19 @@ struct mtk_rx_ring {
 	struct xdp_rxq_info xdp_q;
 };
 
+/* struct mtk_fq_ring -	This struct holds info describing a FQ ring
+ * @scratch_ring:	Newer SoCs need memory for a second HW managed TX ring
+ * @phy_scratch_ring:	physical address of scratch_ring
+ * @scratch_head:	The scratch memory that scratch_ring points to.
+ * @in_sram:		The scratch ring in the SRAM
+ */
+struct mtk_fq_ring {
+	dma_addr_t	phy_scratch_ring;
+	void		*scratch_ring;
+	void		*scratch_head[MTK_FQ_DMA_HEAD];
+	bool		in_sram;
+};
+
 /* struct mtk_rss_params -	This is the structure holding parameters
 				for the RSS ring
  * @hash_key			The element is used to record the
@@ -1337,7 +1350,7 @@ enum mkt_eth_capabilities {
 #define MT7987_CAPS  (MTK_36BIT_DMA | MTK_GMAC1_SGMII | \
 		      MTK_GMAC2_2P5GPHY_V2 | MTK_GMAC2_SGMII | MTK_GMAC3_SGMII | \
 		      MTK_MUX_GMAC123_TO_GEPHY_SGMII | MTK_MUX_GMAC2_TO_2P5GPHY | \
-		      MTK_MUX_U3_GMAC23_TO_QPHY | MTK_U3_COPHY_V2 | \
+		      MTK_MUX_U3_GMAC23_TO_QPHY | MTK_U3_COPHY_V2 | MTK_SRAM | \
 		      MTK_QDMA | MTK_PDMA_INT | MTK_RSS | MTK_HWLRO | \
 		      MTK_GLO_MEM_ACCESS | MTK_RSTCTRL_PPE1)
 
@@ -1517,9 +1530,6 @@ struct mtk_soc_data {
  * @tx_packets:		Net DIM TX packet counter
  * @tx_bytes:		Net DIM TX byte counter
  * @tx_dim:		Net DIM TX context
- * @scratch_ring:	Newer SoCs need memory for a second HW managed TX ring
- * @phy_scratch_ring:	physical address of scratch_ring
- * @scratch_head:	The scratch memory that scratch_ring points to.
  * @clks:		clock array for all clocks required
  * @mii_bus:		If there is a bus we need to create an instance for it
  * @pending_work:	The workqueue used to reset the dma ring
@@ -1554,13 +1564,11 @@ struct mtk_eth {
 	struct mtk_tx_ring		tx_ring;
 	struct mtk_rx_ring		rx_ring[MTK_MAX_RX_RING_NUM];
 	struct mtk_rx_ring		rx_ring_qdma;
+	struct mtk_fq_ring		fq_ring;
 	struct napi_struct		tx_napi;
 	struct mtk_napi			rx_napi[MTK_RX_NAPI_NUM];
 	struct mtk_rss_params		rss_params;
 	struct mtk_qdma_params		qdma_params;
-	void				*scratch_ring;
-	dma_addr_t			phy_scratch_ring;
-	void				*scratch_head[MTK_FQ_DMA_HEAD];
 	struct clk			*clks[MTK_CLK_MAX];
 
 	struct mii_bus			*mii_bus;
